File: selenium_scraper.py
import time
import pandas as pd
import csv
import random
import string
from bs4 import BeautifulSoup
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.keys import Keys

# As with our R script, we will look to collect a range of fuel prices from a price comparison website into a single dataframe

# As my chosen browser is chrome 79, I must match this to chromedriver 79.0.3495
driver = webdriver.Chrome(executable_path=r"/Path/to/chromedriver")

#  We can also initialise browser in headless mode to speed up the process and save on memory
chrome_options = Options()
chrome_options.add_argument("--headless")
driver = webdriver.Chrome(options=chrome_options, executable_path=r"/Users/path_to/chromedriver")

# Custom function we will use to imitate slow typing and fool anti-bot software
def slow_typing(element, text):
    for character in text:
        element.send_keys(character)
        time.sleep(0.3)

# When first visiting the website, we will be asked to register with a name and email address
# As this anti-scraping measure does not require email address confirmation, we will enter in false details to proceed
driver.get("https://www.allstarcard.co.uk/tools/uk-fuel-prices/")

# We will find the 'Name' text box and enter in the following keys
element = driver.find_element_by_css_selector("#Name")
element.send_keys('MyName')

# Next we find the 'Email' text box. As each email only allows a limited number of searches, we will need to create a new one everytime
# We will generate a random string and enter it in email format using string interpolation
def randomString(stringLength=8):
    letters = string.ascii_lowercase
    return ''.join(random.choice(letters) for i in range(stringLength))

email = randomString(8)

element = driver.find_element_by_css_selector("#Email")
element.send_keys(f'{email}@gmail.com')
# We now send the return key from the keyboard to initiate our search
element.send_keys(Keys.RETURN)
# Allow a brief pause for the web page to load
time.sleep(5)

# We are now presented with a new selection box where we can enter a location and view local fuel prices
element = driver.find_element_by_css_selector("#NearbySitesLocation")
element.send_keys('Edinburgh')
element.send_keys(Keys.RETURN)

# A new html table is rendered with our results. We can use BeautifulSoup to find all code with the table class
soup = BeautifulSoup(driver.page_source, 'lxml')
table = soup.find_all('table')[2] # There are several tables within the html so we must grab the correct one

# Handily, pandas contains a function for converting bs4 objects into pd dataframes
df = pd.read_html(str(table))[0]

# We will now remove unnecessary blank columns and add one to show the date of the search and area
df = df.drop(['Map', 'Diesel', 'Petrol'], axis = 1)
df['Search Date'] = pd.to_datetime('today')
df['Area'] = 'Ediburgh'

# This is all we need to collect a dataframe of our fuel prices
# If we wished to search a list of locations and collate the results, we would use a custom function to loop through our list

# Enter in our chosen list of locations to search
location_list = ("Edinburgh", "Orkney", "Glasgow", "Aberdeen", "Girvan", "Kinnaird",
                   "Stranrae", "Moffat", "Aberdeen", "Ibrox")

list_of_dataframes = []
for i in location_list:
    # The registration details are only required once per active session so we need not include this in every loop as it will result in errors
    driver.get("https://www.allstarcard.co.uk/tools/uk-fuel-prices/")

    element = driver.find_element_by_css_selector("#NearbySitesLocation")
    element.send_keys(slow_typing(i))
    element.send_keys(Keys.RETURN)

    soup = BeautifulSoup(driver.page_source, 'lxml')
    table = soup.find_all('table')[2]  # There are several tables within the html so we must grab the correct one

    df = pd.read_html(str(table))[0]
    df = df.drop(['Map', 'Diesel', 'Petrol'], axis=1)
    df['Search Date'] = pd.to_datetime('today')
    df['Area'] = i

    list_of_dataframes.append(df)

final_table = pd.concat(list_of_dataframes)

File: selenium_scraper.R
# Import the necessary libraries
library(openxlsx)
library(stringr)
library(purrr)
library(RSelenium)
library(readr)
library(dplyr)
library(glue)
library(rvest)

# We will use selenium to navigate to a fuel prices comparison website and collect data to compare prices across Scotland
# This website has registration barriers and reactive javascript but we will show how selenium can easily bypass this

# Open browser using selenium after installing the necessary software
driver <- rsDriver(
  browser = "firefox", geckover = "latest",
  extraCapabilities = list(
    "moz:firefoxOptions" = list(
      args = list('--headless')
    )
  )
)

# Note if you receive a selenium error at this point, you most likely have to update your firefox gecko driver
# Run binman::list_versions("geckodriver") to see available drivers
# Use following code to change RSelenium default version used to compatible version
# driver <- rsDriver(browser=c("firefox"), geckover="8.0.3904.70") or simply put 'latest' if your drivers are all up to date

# This is a headless version but, if you would like to see what your code is doing, you can use:
driver <- rsDriver(
  browser = "firefox", geckover = "latest"
)

remDr <- driver[["client"]]

# Open URL
remDr$navigate("https://www.allstarcard.co.uk/tools/uk-fuel-prices/")

# Allow a pause for the html to load
Sys.sleep(3)

# Now we can see that we must register in order to proceed but testing shows that the user simply needs to submit details, not reply to an email registration
# So we will identify the input boxes using css selector and enter faux registration details
Name <- remDr$findElement(using = 'css selector',  "#Name")
Name$sendKeysToElement(list('MyName'))

# Now for each email registration, only a set number of searches are allowed so we will need to enter a new false email with each scrape
# Generate a random email address using a random string generator and string interpolation
random_string_generator <- function(n = 5000) {
  a <- do.call(paste0, replicate(5, sample(LETTERS, n, TRUE), FALSE))
  paste0(a, sprintf("%04d", sample(9999, n, TRUE)), sample(LETTERS, n, TRUE))
}

email <- random_string_generator(1)

Email <- remDr$findElement(using = 'css selector',  "#Email")
Email$sendKeysToElement(list(glue('{email}@gmal.com'), key = "enter")) # Hit enter and submit our details 

# We can now see that the webpage is ready for us to submit our locations. Again, find the necessary box using chrome css selector
searchBox <- remDr$findElement(using = 'css selector',  "#NearbySitesLocation")
  searchBox$sendKeysToElement(list(location, key = "enter"))

# But we want to look at a list of locations so will create a custom function that will enter in each location and collect the resulting table

ScrapeFuel <- function(driver, location) {
  
  remDr <- driver[["client"]]
  
  # Open URL again - necessary as there is no search box in the table render page
  remDr$navigate("https://www.allstarcard.co.uk/tools/uk-fuel-prices/")
  
  # Type in location to input text box
  searchBox <- remDr$findElement(using = 'css selector',  "#NearbySitesLocation")
  searchBox$sendKeysToElement(list(location, key = "enter"))
  
  # Get our data by passing the url object to rvest and collecting using html_nodes 
  html <- remDr$getPageSource()[[1]]
  
  table <- read_html(html) %>%
    html_nodes("table.table.table-bordered.results-table") %>% # we find the table class name using web developer tools in chrome
    .[[1]] %>%
    html_table(fill=T) #html_table handily converts data packets into data frame format
  
  # Add additional columns and tidy away empty ones
  table$Date <- Sys.Date()
  table$Location <- location
  table$Map <- NULL
  table$Petrol <- NULL
  table$Diesel <- NULL
  
  return(table)
  
}

# So now we have our custom function, let's put together a list of locations we want it to search for
location_list <- c("Dunrossness", "Orkney", "Aberfeldy", "Kinnaird", 
                   "Springside", "Moffat", "Linton", "Ibrox", "Edinburgh", "Glasgow", 
                   "Dundee", "Aberdeen", "Stranrae", "Girvan", "Fife")

# and using assign and lapply we apply this custom function to every location and put the resulting df into a list
fuel_prices_list <- lapply(location_list, function(i) {
  assign(i, ScrapeFuel(driver, i))
})

# Use to close browser
driver$server$stop()

# Bundle all our dataframes together with rbind and we have our comprehensive list of fuel prices across Scotland
fuel_prices_current <- bind_rows(fuel_prices_list)

# So the complete workflow will be:
# 1. Naviagate to url and create registration details
# 2. Cycle through list of location searches using custom function
# 3. Collect and collate results into a single dataframe

# END OF SCRIPT
